### PR TITLE
Added a ComplexConcat layer

### PR DESCRIPTION
**Pull request by matt-gardner**

### CODE REVIEW

File: src/main/python/deep_qa/layers/complex_concat.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        can do elementwise operations on them, and all tensors must have the same number of
+        dimensions, and match on all dimensions except the concatenation axis.
+    Output: A tensor with some combination of the input tensors concatenated along a specific
+        dimension.
+    Parameters:
+        - `axis`: the axis to use for K.concatenate
+        - `combination`: a comma-separated list of combinations to perform on the input tensors.
+          These are either tensor indices (1-indexed), or an arithmetic operation between two
+          tensor indices.  For example, these are all valid combination parameters: "1,2", "1,2*3",
+          "1-2,2-1", and "1,2,1*2".
```
Comment (nelson-liu) [inline src/main/python/deep_qa/layers/complex_concat.py]: maybe mention what the valid arithmetic operations are?

File: src/main/python/deep_qa/layers/complex_concat.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        super(ComplexConcat, self).__init__(**kwargs)
+
+    @overrides
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        return mask[0]
+
+    @overrides
+    def get_output_shape_for(self, input_shape):
+        assert isinstance(input_shape, list), "ComplexConcat must have multiple inputs"
```
Comment (nelson-liu) [inline src/main/python/deep_qa/layers/complex_concat.py]: maybe verify that the list has more than 1 element as well?

Comment (matt-gardner) [inline src/main/python/deep_qa/layers/complex_concat.py]: This combination string is also valid: "1,1*1".

File: doc/api_doc/deep_qa.layers.rst [hunk 3]
```diff
@@ -23,6 +23,14 @@ deep_qa.layers.additive
     :undoc-members:
     :show-inheritance:
 
+deep_qa.layers.complex_concat
+-----------------------
```
Comment (matt-gardner) [inline doc/api_doc/deep_qa.layers.rst]: Done.

File: src/main/python/deep_qa/layers/complex_concat.py (hunk 1 above)
Comment (matt-gardner) [inline src/main/python/deep_qa/layers/complex_concat.py]: Done.

File: src/main/python/deep_qa/layers/complex_concat.py (hunk 2 above)
Comment (matt-gardner) [inline src/main/python/deep_qa/layers/complex_concat.py]: Done.

**Pull request merged**